### PR TITLE
refactor(client/challenge views): extract items into sharable components

### PR DESCRIPTION
**Pull request by moT01**
Many of the view components for our challenges have duplicate features - e.g. multiple choice questions, assignments, and other things. This extracts the assignments, mc questions, and fill in the blanks into their own components - so we can reuse them in multiple places.

This will make it easier to add the many mc questions per challenge that I am trying to add.

There shouldn't be much for visible changes. A few minor things - mainly, it moves the "Please complete the assignments" to right below the assignments.

Also, if I get around to it - the end goal might be to have a single view component that incorporates all these features.

Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #XXXXX

### CODE REVIEW

**Comment by Sembauke** [comment]
I would love this.

File: client/src/templates/Challenges/components/assignments.tsx [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+      <ChallengeHeading heading={t('learn.assignments')} />
+      <div className='video-quiz-options'>
+        {assignments.map((assignment, index) => (
+          <label className='video-quiz-option-label' key={index}>
+            <input
+              name='assignment'
+              type='checkbox'
+              onChange={event =>
+                handleAssignmentChange(event, assignments.length)
+              }
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/components/assignments.tsx:36]: False positive, we can ignore this.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]

File: client/src/templates/Challenges/components/assignments.tsx [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+function Assignments({
+  assignments,
+  allAssignmentsCompleted,
+  handleAssignmentChange
+}: AssignmentsProps): JSX.Element {
+  const { t } = useTranslation();
+  return (
+    <>
+      <ChallengeHeading heading={t('learn.assignments')} />
```
[the remark this answers is not in the document]
Comment (moT01) [reply client/src/templates/Challenges/components/assignments.tsx:27]: This was pre-existing, so I think it's a little outside the scope of this PR - I would prefer to leave it here. I plan on some more refactoring of these, so I can potentially try to change it in another PR. Also, I don't know if I like either of those any better - what about `featureHeading` or `challengeTaskHeading` or something? To be clear, they are these headings in the challenges:

<details><summary>image</summary>

Screenshot 2024-09-03 at 8 26 07 AM

</details>

File: client/src/templates/Challenges/fill-in-the-blank/show.tsx [hunk 3]
```diff
@@ -237,25 +236,6 @@ class ShowFillInTheBlank extends Component<
...
-  addCodeTags(str: string, index: number, numberOfBlanks: number): string {
-    if (index === 0) return `${str}</code>`;
-    if (index < numberOfBlanks) return `<code>${str}</code>`;
-    return `<code>${str}`;
-  }
-
-  addPrismClass(index: number, numberOfBlanks: number): string {
-    if (index === 0) return `first-code-tag`;
-    if (index < numberOfBlanks) return `middle-code-tag`;
-    return `last-code-tag`;
```
[the remark this answers is not in the document]
Comment (moT01) [reply client/src/templates/Challenges/fill-in-the-blank/show.tsx:249]: ✅

File: client/src/templates/Challenges/components/fill-in-the-blanks.tsx [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+                      type='text'
+                      maxLength={
+                        blankAnswers[value as unknown as number].length + 3
+                      }
+                      className={`fill-in-the-blank-input ${addInputClass(
+                        value as unknown as number
+                      )}`}
+                      onChange={handleInputChange}
+                      data-index={node.value}
+                      size={blankAnswers[value as unknown as number].length}
```
[the remark this answers is not in the document]
Comment (moT01) [reply client/src/templates/Challenges/components/fill-in-the-blanks.tsx]: ✅

File: client/src/templates/Challenges/components/fill-in-the-blanks.tsx [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+                if (type === 'blank')
+                  return (
+                    <input
+                      key={j}
+                      type='text'
+                      maxLength={
+                        blankAnswers[value as unknown as number].length + 3
+                      }
+                      className={`fill-in-the-blank-input ${addInputClass(
+                        value as unknown as number
```
[the remark this answers is not in the document]
Comment (moT01) [reply client/src/templates/Challenges/components/fill-in-the-blanks.tsx]: ✅

File: client/src/templates/Challenges/components/fill-in-the-blanks.tsx [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+                const { type, value } = node;
+                if (type === 'text') return value;
+                if (type === 'blank')
+                  return (
+                    <input
+                      key={j}
+                      type='text'
+                      maxLength={
+                        blankAnswers[value as unknown as number].length + 3
+                      }
```
[the remark this answers is not in the document]
Comment (moT01) [reply client/src/templates/Challenges/components/fill-in-the-blanks.tsx]: ✅

File: client/src/templates/Challenges/components/assignments.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (ojeytonwilliams) [reply client/src/templates/Challenges/components/assignments.tsx:27]: No worries, it's not urgent.

As for what to call them, it's just that they don't have anything to do with challenges beyond the fact they're used there. Not a big deal, obviously.

**Review by ojeytonwilliams — APPROVED** [review]
LGTM 👍

**Review by Sembauke — APPROVED** [review]

**Pull request merged**